### PR TITLE
Initial PyTorch Dataset support

### LINKED ISSUE #109: Support reading lance into PyTorch
We can start to use lance for training and inference.

### PR DESCRIPTION
**Pull request by eddyxu**
Basic and experimental implementation of `torch.Dataset`. 

Closes #109

### CODE REVIEW

File: python/lance/_lib.pyx [hunk 1]
```diff
@@ -4,13 +4,22 @@ from typing import Optional, Union
...
 from libcpp cimport bool
-from libcpp.memory cimport shared_ptr, const_pointer_cast
+from libcpp.memory cimport const_pointer_cast, shared_ptr
 from libcpp.string cimport string
 
 from pathlib import Path
 
 from pyarrow import Table
-from pyarrow._dataset cimport FileFormat, FileWriteOptions, CFileWriteOptions, CScanner, CDataset, Dataset
+
```
Comment (eddyxu) [inline python/lance/_lib.pyx:13]: These changes were due to running `isort`

File: python/lance/__init__.py [hunk 2]
```diff
@@ -49,13 +49,19 @@ def scanner(
     data: Union[str, Path, ds.Dataset],
     columns: Optional[str] = None,
     filter: Optional[pc.Expression] = None,
+    batch_size: Optional[int] = None,
```
Comment (changhiskhan) [inline python/lance/__init__.py:52]: Let's add docstrings for the function?

File: python/setup.py [hunk 3]
```diff
@@ -69,7 +67,8 @@
     ext_modules=cythonize(extensions, language_level="3"),
     zip_safe=False,
     install_requires=["pyarrow>=9,<10"],
-    extras_require={"test": ["pytest>=6.0", "pandas", "duckdb", "click"]},
+    extras_require={"test": ["pytest>=6.0", "pandas", "duckdb", "click"],
+                    "pytorch": ["torch"]},
```
Comment (changhiskhan) [inline python/setup.py:71]: I'm wondering if we should just ask the user to manually pip install torch? I feel like Rikai's extras system is way too complicated and hard to maintain. I'm wondering whether we should keep it simple for now.

File: python/lance/tests/test_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+
+pytest.importorskip("torch")
```
Comment (changhiskhan) [inline python/lance/tests/test_pytorch.py:17]: torch = pytest.importorskip("torch")? Save the import torch call afterwards

File: python/lance/tests/test_api.py [hunk 5]
```diff
@@ -16,10 +16,10 @@
 from pathlib import Path
```
Comment (changhiskhan) [inline python/lance/tests/test_api.py:16]: Changes in this whole file is just formatting right?

Comment (eddyxu) [reply python/lance/tests/test_api.py:16]: yes, that's right.

File: python/lance/tests/test_pytorch.py (hunk 4 above)
Comment (eddyxu) [reply python/lance/tests/test_pytorch.py:17]: Oh, i dont know this works. TIL

File: python/setup.py (hunk 3 above)
Comment (eddyxu) [reply python/setup.py:71]: Hmm ok.

File: python/lance/__init__.py (hunk 2 above)
Comment (eddyxu) [reply python/lance/__init__.py:52]: Done

File: .github/workflows/python.yml [hunk 6]
```diff
@@ -31,7 +31,7 @@ jobs:
       run: ./tools/build_wheel.sh $(echo cp${{ matrix.python-version }} | sed "s/\.//")
     - name: Pip install
       run: |
-        python -m pip install $(ls wheels/*.whl)[test]
+        python -m pip install $(ls wheels/*.whl)[test,pytorch]
```
Comment (changhiskhan) [inline .github/workflows/python.yml:34]: install pytorch directly?

**Review by changhiskhan — APPROVED** [review]
the gh action needs update but lgtm otherwise

File: .github/workflows/python.yml (hunk 6 above)
Comment (eddyxu) [reply .github/workflows/python.yml]: Fixed.

**Pull request merged**